corehq/motech/repeaters/tests/test_payload_generators.py

```diff
@@ -0,0 +1,184 @@
+
+        self.assertIsInstance(form_dict['form']['decimal'], str)
+        self.assertIsInstance(info.form_question_values['/data/decimal'], str)
+        self.assertIsInstance(info.updates['decimal'], str)
+        self.assertIsInstance(info.extra_fields['decimal'], Decimal)  # <--
+
+    def test_form_json_integer(self):
+        """
+        When an integer is taken from form JSON or a case block, it is a
+        ``str``. When it is taken from a case property, it is still a ``str``.
+        """
+        gen = FormRepeaterJsonPayloadGenerator(None)
+        form_dict, info = self.get_payload_info(gen)
+
+        self.assertIsInstance(form_dict['form']['year'], str)
+        self.assertIsInstance(info.form_question_values['/data/year'], str)
+        self.assertIsInstance(info.updates['year'], str)
+        self.assertIsInstance(info.extra_fields['year'], str)  # <--
+
+    def test_form_dict_integer(self):
+        """
+        When an integer is taken from form JSON or a case block, it is a
+        ``str``. When it is taken from a case property, it is still a ``str``.
+        """
+        gen = FormDictPayloadGenerator(None)
+        form_dict, info = self.get_payload_info(gen)
+
+        self.assertIsInstance(form_dict['form']['year'], str)
+        self.assertIsInstance(info.form_question_values['/data/year'], str)
+        self.assertIsInstance(info.updates['year'], str)
+        self.assertIsInstance(info.extra_fields['year'], str)  # <--
+
+    def test_form_json_multiplechoice(self):
+        gen = FormRepeaterJsonPayloadGenerator(None)
+        form_dict, info = self.get_payload_info(gen)
+
+        self.assertIsInstance(form_dict['form']['breakfast'], str)
+        self.assertIsInstance(info.form_question_values['/data/breakfast'], str)
+        self.assertIsInstance(info.updates['breakfast'], str)
+        self.assertIsInstance(info.extra_fields['breakfast'], str)
+
+    def test_form_dict_multiplechoice(self):
+        gen = FormDictPayloadGenerator(None)
+        form_dict, info = self.get_payload_info(gen)
+
+        self.assertIsInstance(form_dict['form']['breakfast'], str)
+        self.assertIsInstance(info.form_question_values['/data/breakfast'], str)
+        self.assertIsInstance(info.updates['breakfast'], str)
+        self.assertIsInstance(info.extra_fields['breakfast'], str)
+
+    def get_payload_info(
+        self,
+        payload_generator: BasePayloadGenerator,
+    ) -> Tuple[dict, CaseTriggerInfo]:
+        payload = payload_generator.get_payload(None, self.form)
+        if isinstance(payload, str):
+            payload = json.loads(payload)
+        [info] = get_relevant_case_updates_from_form_json(
+            DOMAIN,
+            payload,
+            case_types=None,
+            extra_fields=['year', 'breakfast', 'decimal'],
+            form_question_values=get_form_question_values(payload),
+        )
+        return payload, info
+
+
+def post_xform(xform_id, case_id):
```

Comment:
You could use `CaseFactory` here.

Reply:
:grin: Ha! Thank you.

Reply:
Ended up using FormSubmissionBuilder instead of CaseFactory because it makes it easier to use the form. Added tests for SQL domains. Unexpected (by me) behaviour only happens in Couch.